src/passes/inlineAssemblyTransformer.ts

```diff
@@ -0,0 +1,157 @@
+import { ASTMapper } from '../ast/mapper';
+import { AST } from '../ast/ast';
+import {
+  createExpressionStatement,
+  createIdentifier,
+  createNumberLiteral,
+  createTuple,
+} from '../utils/nodeTemplates';
+
+import {
+  InlineAssembly,
+  YulNode,
+  Identifier,
+  ASTNode,
+  BinaryOperation,
+  Assignment,
+  Block,
+  Expression,
+  ExpressionStatement,
+  VariableDeclaration,
+  UncheckedBlock,
+  Literal,
+  Statement,
+} from 'solc-typed-ast';
+import assert from 'assert';
+import { getErrorMessage, WillNotSupportError } from '../utils/errors';
+
+const binaryOps: { [name: string]: string } = { add: '+', mul: '*', div: '/', sub: '-' };
+
+class YulVerifier {
+  errors: string[] = [];
+
+  verifyNode(node: YulNode) {
+    const methodName = `check${node.nodeType}`;
+    if (!(methodName in YulTransformer.prototype))
+      this.errors.push(`${node.nodeType} is not supported`);
+    const method = this[methodName as keyof YulVerifier] as (node: YulNode) => void;
+    if (method !== undefined) return method.bind(this)(node);
+  }
+  checkYulAssignment(node: YulNode) {
+    this.verifyNode(node.value);
+  }
+
+  checkYulFunctionCall(node: YulNode) {
+    if (binaryOps[node.functionName.name] === undefined)
+      this.errors.push(`${node.functionName.name} is not supported`);
+    node.arguments.forEach((arg: YulNode) => this.verifyNode(arg));
+  }
+}
+
+class YulTransformer {
+  vars: Map<string, VariableDeclaration>;
+  ast: AST;
+
+  constructor(ast: AST, assemblyRoot: InlineAssembly) {
+    this.ast = ast;
+    const blockRefIds = assemblyRoot.externalReferences.map((ref) => ref.declaration);
+    assert(assemblyRoot.context !== undefined, `Assembly root context not found.`);
+    const blockRefs = [...assemblyRoot.context.map].filter(([id, _]) =>
+      blockRefIds.includes(id),
+    ) as [number, VariableDeclaration][];
+    this.vars = new Map(blockRefs.map(([_, ref]) => [ref.name, ref]));
+  }
+
+  toSolidityStatement(node: YulNode): Statement {
```

Comment:
I've used an explicit naming. It could be prefixed with `build` or `convert`

Reply:
No issue here.